Sourcery/Parsing/Composer.swift

```diff
@@ -91,13 +91,74 @@ struct Composer {
         return unique.values.sorted { $0.name < $1.name }
     }
 
+    private func resolveGenerics(typeName: TypeName, unique: [String: Type]) {
+        let genericSplit = typeName.unwrappedTypeName.characters.split(separator: "<", maxSplits: 1)
+
+        guard genericSplit.count == 2 else {
+            return
+        }
+
+        let typeCharacters = genericSplit[0]
+        var genericCharacters = genericSplit[1]
+
+        guard genericCharacters.count > 1, genericCharacters.removeLast() == ">" else {
+            return
+        }
+
+        var genericTypeStrings = [String]()
+
+        var subGenericsCounter = 0
+        var characterBuffer = [Character]()
+
+        for character in genericCharacters {
+            switch character {
+            case "<":
+                characterBuffer.append(character)
+                subGenericsCounter += 1
+            case ">":
+                characterBuffer.append(character)
+                subGenericsCounter -= 1
+            case ",":
+                if subGenericsCounter == 0 {
+                    genericTypeStrings.append(String(characterBuffer).trimmingCharacters(in: [" "]))
+                    characterBuffer = []
+                } else {
+                    characterBuffer.append(character)
+                }
+            default:
+                characterBuffer.append(character)
+            }
+        }
+
+        if !characterBuffer.isEmpty {
+            genericTypeStrings.append(String(characterBuffer).trimmingCharacters(in: [" "]))
+        }
+
+        let genericTypeNames = genericTypeStrings.flatMap { genericTypeName -> TypeName in
+            let typeName = TypeName(genericTypeName)
+            resolveGenerics(typeName: typeName, unique: unique)
+            return typeName
+        }
+
+        let genericTypes = genericTypeStrings.flatMap { genericTypeName in
+            return unique[genericTypeName]
+        }
+
+        typeName.generic = GenericType(name: String(typeCharacters), referencedTypes: genericTypes, referencedTypeNames: genericTypeNames)
+    }
+
     private func resolveType(typeName: TypeName, containingType: Type?, unique: [String: Type], modules: [String: [String: Type]], typealiases: [String: Typealias]) -> Type? {
         let actualTypeName = self.actualTypeName(for: typeName, containingType: containingType, unique: unique, typealiases: typealiases)
         if let actualTypeName = actualTypeName, actualTypeName != typeName.unwrappedTypeName {
             typeName.actualTypeName = TypeName(actualTypeName)
         }
 
         let lookupName = typeName.actualTypeName ?? typeName
+
+        if lookupName.isGeneric, lookupName.generic == nil {
```

Comment:
I wonder if we should actually consider dictionary and array types as generics, given that they are already represented with corresponding properties. The thing that parsing results will be different depending on how these types are defined (the same applies to optionals too btw) - if they are defined with a full syntax the will be parsed both as generic and dictionary/array, if shorthand syntax is used they will be not recognised as generics. Also generic types data is duplicated in this case and additional parsing seems to me as unneeded. And most likely arrays and dictionaries should be handled separately in templates - they can not be always handled as generics because of shorthand syntax case and they are usually treated as a special case of types.

So I would put parsing generics in the very end after parsing type as array, dict and tuple.
What do you think @krzysztofzablocki ?